### ISSUE TITLE
Specifying secrets backend

### ISSUE THREAD

**Issue by dalekurt** [channable/vaultenv#45]
I have created a new secret on vault called `demo` so all my secrets are stored there. 

However, I do notice that vaultenv is using the path `/v1/secret/demo/my/key/value` to find my key values and not `/v1/demo/my/key/value`. Is there away to configure the backend?

```
$ vault write secret/production/third-party api-key=fecb0f6e97c5b37b3a814107682cf68416f072a8
              ^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^ ^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
             backend          path            key                   value
```

**Comment by ruuda** [comment]
Is it useful (or even possible) to use more than one backend at the same time? If not, I would take the backend from a command-line option or environment variable (possibly with a default value) rather than complicating the config file.

**Comment by duijf** [comment]
> Is it useful (or even possible) to use more than one backend at the same time?

Yes and yes :) 

I think Vault now uses new terminology, the page in the docs is called ["Secret Engines"](https://www.vaultproject.io/docs/secrets/index.html). You can for example have a KV engine (previously called a generic backend), and a Database engine.

The first functions like encrypted key/value storage (like we have running in prod at Channable). The second engine has the same API, but generates credentials on the fly by connecting to Postgres/MySQL and running `CREATE USER` statements. There's another one for SSH.

**Comment by dalekurt** [comment]
@duijf Thank you for getting back to me on this "feature request" I do like the approach of defining the "Secret Engine" (or Backend) in the examples above. That would allow for both static and dynamic key values. At the moment with my own implementation I have dynamic key values for mySQL which query a `database/` path on Vault stored to a JSON file then using jq I can isolate and populate the environment variables. Using your proposed method I would be able to define the path within the `.secrets` file. I like that idea.

**Comment by duijf** [comment]
That's something we'd like to support explicitly without users having to glue these things together

I'm currently implementing this. Will keep the config file roughly like I proposed but:

 - I'll change `BACKEND` to `MOUNT`, as that's the new terminology
 - I don't think I will enforce the newlines
 - I'll introduce `VERSION 2` to get better compatibility guarantees in the future

Implementation note: since the parser is becoming more involved, I'll probably make use of some parser library. It's been a while since I played with parser combinators, so I'm going to have to see what's what.

**Comment by costimuraru** [comment]
👍

**Comment by shrinidhi-atmakur** [comment]
@duijf, I am using the latest ( v.0.7.1) version and tried using the MOUNT specify the secret backed, but noticed that the vault path is still prefixed by /v1/secret. Is the above feature request completed?

MOUNT demo/
staging/third-party#api-key

Thank you.

**Comment by duijf** [comment]
Hi @shrinidhi-atmakur! Thanks for your interest. As you can see from this ticket being open, we're not completely ready yet. I'm trying to get the PR https://github.com/channable/vaultenv/pull/47 in today or next Monday and will make a new vaultenv release once it has been merged :)

**Comment by duijf** [comment]
I've merged #47 and am making a new release.

Thanks for the patience everyone :)
